crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -320,8 +320,94 @@ unsafe impl<D: QueryData + 'static, F: QueryFilter + 'static> SystemParam for Qu
         // world data that the query needs.
         unsafe { Query::new(world, state, system_meta.last_run, change_tick) }
     }
+
+    fn apply(state: &mut Self::State, system_meta: &SystemMeta, world: &mut World) {
+        world.init_resource::<ComponentMutationBuffer>();
+        world.resource_scope(|world, mut buffer: Mut<ComponentMutationBuffer>| {
+            let (writes, cant_write) = state.component_access.access.component_writes();
+            let mutable_components = if cant_write { Vec::new() } else { writes.collect() };
+            if mutable_components.is_empty() { return }
+
+            let buffer_started_empty = buffer.0.is_empty();
+    
+            for archetype_id in state.matched_archetypes() {
+                let Some(archetype) = world.archetypes().get(archetype_id) else { continue; };
+                if !archetype.has_mutate_observer() { continue; }
+                
+                for archetype_entity in archetype.entities() {
+                    let entity = archetype_entity.id();
+                    let entity_ref = world.entity(entity);
+                    let mutated_observers = mutable_components
+                        .iter()
+                        .cloned()
+                        .filter(|c| entity_ref
+                            .get_change_ticks_by_id(*c)
+                            .map(|ticks| ticks.changed.get() >= system_meta.last_run.get()) // FIXME GRACE: get this_run
+                            .unwrap_or(false)
+                        ); 
+                    let values = buffer.0.entry(entity).or_default();
+                    values.extend(mutated_observers);
+                }
+            }
+
+            if buffer_started_empty && !buffer.0.is_empty() {
+                world.commands().queue(flush_mutations);
```

Comment:
This is a jank way that I'm deduplicating `flush_mutations` commands, where I'm only adding a `flush_mutations` command to the queue when `ComponentMutationBuffer` starts empty and has something added to it. Is there a better way of doing this?